src/EditorFeatures/Core/Implementation/IntelliSense/Completion/AsyncCompletionService.cs

```diff
@@ -57,31 +68,49 @@ public bool TryGetController(ITextView textView, ITextBuffer subjectBuffer, out
         {
             AssertIsForeground();
 
-            // check whether this feature is on.
-            if (!subjectBuffer.GetFeatureOnOffOption(InternalFeatureOnOffOptions.CompletionSet))
+            if (!IsModernCompletionEnabled(textView, subjectBuffer))
             {
                 controller = null;
                 return false;
             }
 
+            var autobraceCompletionCharSet = GetAllAutoBraceCompletionChars(subjectBuffer.ContentType);
+            controller = Controller.GetInstance(
+                ThreadingContext,
+                textView, subjectBuffer,
+                _editorOperationsFactoryService, _undoHistoryRegistry, _completionPresenter,
+                _listener,
+                autobraceCompletionCharSet);
+
+            return true;
+        }
+
+        private bool IsModernCompletionEnabled(ITextView textView, ITextBuffer subjectBuffer)
+        {
+            if (!_completionAPIEnabled.HasValue)
+            {
+                if (Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out var workspace))
+                {
+                    var experimentationService = workspace.Services.GetService<IExperimentationService>();
+                    _completionAPIEnabled = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.CompletionAPI);
+                }
+            }
+
+            // check whether the feature flag (async completion API) is not set and this feature is on.
+            if (_completionAPIEnabled == true || !subjectBuffer.GetFeatureOnOffOption(InternalFeatureOnOffOptions.CompletionSet))
```

Comment:
Are we checking this FeatureOnOffOptions flag anywhere in the new implementation?

Reply:
No, we do not. If a new implementation is called, we assume the editor uses the new API.

---
In reply to: [221396073](https://github.com/dotnet/roslyn/pull/29016#discussion_r221396073) [](ancestors = 221396073)